contrib/devtools/bitcoin-tidy/nontrivial-threadlocal.cpp

```diff
@@ -0,0 +1,44 @@
+// Copyright (c) 2023 Bitcoin Developers
+// Distributed under the MIT software license, see the accompanying
+// file COPYING or http://www.opensource.org/licenses/mit-license.php.
+
+#include "nontrivial-threadlocal.h"
+
+#include <clang/AST/ASTContext.h>
+#include <clang/ASTMatchers/ASTMatchFinder.h>
+
+
+// Copied from clang-tidy's bugprone check
+namespace {
+AST_MATCHER(clang::CXXRecordDecl, hasNonTrivialDestructor) {
+  // TODO: If the dtor is there but empty we don't want to warn either.
```

Comment:
How about:
```diff
diff --git a/contrib/devtools/bitcoin-tidy/example_nontrivial-threadlocal.cpp b/contrib/devtools/bitcoin-tidy/example_nontrivial-threadlocal.cpp
index 2b74df5d0e..b002712baf 100644
--- a/contrib/devtools/bitcoin-tidy/example_nontrivial-threadlocal.cpp
+++ b/contrib/devtools/bitcoin-tidy/example_nontrivial-threadlocal.cpp
@@ -2,0 +3,30 @@ thread_local std::string foo;
+thread_local char* bar;
+
+struct Test_Ok_1 {
+    int member;
+    Test_Ok_1() {
+        member = 1;
+    }
+    ~Test_Ok_1() {}
+};
+
+struct Test_Ok_2 {
+    int member;
+    Test_Ok_2() {
+        member = 1;
+    }
+};
+
+struct Test_Bad {
+    int member;
+    Test_Bad() {
+        member = 1;
+    }
+    ~Test_Bad() {
+        member = 0;
+    }
+};
+
+thread_local Test_Ok_1 test_1;
+thread_local Test_Ok_2 test_2;
+thread_local Test_Bad test_3;
diff --git a/contrib/devtools/bitcoin-tidy/nontrivial-threadlocal.cpp b/contrib/devtools/bitcoin-tidy/nontrivial-threadlocal.cpp
index bce1a99165..3f5a85e444 100644
--- a/contrib/devtools/bitcoin-tidy/nontrivial-threadlocal.cpp
+++ b/contrib/devtools/bitcoin-tidy/nontrivial-threadlocal.cpp
@@ -14,2 +14,19 @@ AST_MATCHER(clang::CXXRecordDecl, hasNonTrivialDestructor) {
-  // TODO: If the dtor is there but empty we don't want to warn either.
-  return Node.hasDefinition() && Node.hasNonTrivialDestructor();
+  if (!Node.hasDefinition()) {
+    return false;
+  }
+  if (Node.hasTrivialDestructor()) {
+    return false;
+  }
+  const clang::CXXDestructorDecl* dtor = Node.getDestructor();
+  if (!dtor) {
+    return false;
+  }
+
+  if (dtor->hasBody()) {
+    const clang::Stmt* body = dtor->getBody();
+    if (const clang::CompoundStmt* comp_stmt = clang::dyn_cast<clang::CompoundStmt>(body)) {
+      return !comp_stmt->body_empty();
+    }
+  }
+
+  return false;
```

Reply:
@TheCharlatan Thanks for playing around with it :)

I don't have the internal clang knowledge to know if that's correct or not. Looks fine to me, but I agree with @maflcko that we could just remove the empty dtor in that case.

Also, to be clear, that TODO is an upstream comment, not one from me: https://github.com/llvm/llvm-project/blob/main/clang-tools-extra/clang-tidy/bugprone/UnusedRaiiCheck.cpp#L18

Reply:
> I don't have freeBSD to test this, but are you sure it doesn't crash when the dtor is there but empty?

It looks like it should work to me. If a method has a body it should also contain a compound statement, or at least this [book](https://github.com/PacktPublishing/LLVM-Techniques-Tips-and-Best-Practices-Clang-and-Middle-End-Libraries) seems to suggest as much.

> Also, to be clear, that TODO is an upstream comment, not one from me: https://github.com/llvm/llvm-project/blob/main/clang-tools-extra/clang-tidy/bugprone/UnusedRaiiCheck.cpp#L18

...but if it is an upstream TODO I might well be missing something.